dpctl/apis/include/dpctl4pybind11.hpp

```diff
@@ -748,6 +748,64 @@ class usm_memory : public py::object
             throw py::error_already_set();
     }
 
+    /*! @brief Create usm_memory object from shared pointer that manages
+     *  lifetime of the USM allocation.
+     */
+    usm_memory(void *usm_ptr,
+               size_t nbytes,
+               const sycl::queue &q,
+               std::shared_ptr<void> shptr)
+    {
+        auto const &api = ::dpctl::detail::dpctl_capi::get();
+        DPCTLSyclUSMRef usm_ref = reinterpret_cast<DPCTLSyclUSMRef>(usm_ptr);
+        sycl::queue *q_ptr = new sycl::queue(q);
```

Comment:
We need to use shared pointer because it is used by Python host code, and by host tasks to ensure life-time of USM allocation extends past completion of kernel execution submissions.

Reply:
What I was talking about is to use `unique_ptr` to manage created objects and free it in case of failure automatically:
```C++
auto q_ptr = std::make_unique(new sycl::queue(q));
DPCTLSyclQueueRef QRef = reinterpret_cast<DPCTLSyclQueueRef>(q_ptr.get());
...

if (is_ok) {
    ...
    m_ptr = _memory;
    q_ptr.release();
}
```
So we don't need to call
```C++
delete q_ptr;
```
by hands. And memory would be freed in any scenario except one where we are calling `release`

Reply:
Thank you @AlexanderKalistratov ! I like the suggestion. I am going to push the change shortly